File: public/public-persistence-jooq/src/main/java/ch/difty/scipamato/publ/persistence/code/JooqCodeRepo.java
package ch.difty.scipamato.publ.persistence.code;

import static ch.difty.scipamato.common.persistence.TranslationUtilsKt.trimLanguageCode;
import static ch.difty.scipamato.publ.db.tables.Code.CODE;

import java.util.List;

import org.jetbrains.annotations.NotNull;
import org.jooq.DSLContext;
import org.springframework.cache.annotation.CacheConfig;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Repository;

import ch.difty.scipamato.common.entity.CodeClassId;
import ch.difty.scipamato.publ.entity.Code;

@Repository
@CacheConfig(cacheNames = "codes")
public class JooqCodeRepo implements CodeRepository {

    private final DSLContext dslContext;

    public JooqCodeRepo(@NotNull final DSLContext dslContext) {
        this.dslContext = dslContext;
    }

    @NotNull
    @Override
    @Cacheable
    public List<Code> findCodesOfClass(@NotNull final CodeClassId codeClassId, @NotNull final String languageCode) {
        final String lang = trimLanguageCode(languageCode);
        // skipping the audit fields
        return dslContext
            .select(CODE.CODE_CLASS_ID, CODE.CODE_, CODE.LANG_CODE, CODE.NAME, CODE.COMMENT, CODE.SORT)
            .from(CODE)
            .where(CODE.LANG_CODE
                .eq(lang)
                .and(CODE.CODE_CLASS_ID.equal(codeClassId.getId())))
            .orderBy(CODE.CODE_CLASS_ID, CODE.SORT)
            .fetchInto(Code.class);
    }
}

File: public/public-persistence-jooq/src/main/kotlin/ch/difty/scipamato/publ/persistence/code/JooqCodeRepo.kt
package ch.difty.scipamato.publ.persistence.code

import ch.difty.scipamato.common.entity.CodeClassId
import ch.difty.scipamato.common.persistence.trimLanguageCode
import ch.difty.scipamato.publ.entity.Code
import org.jooq.DSLContext
import org.springframework.cache.annotation.CacheConfig
import org.springframework.cache.annotation.Cacheable
import org.springframework.stereotype.Repository

@Repository
@CacheConfig(cacheNames = ["codes"])
open class JooqCodeRepo(private val dslContext: DSLContext) : CodeRepository {

    @Suppress("MagicNumber")
    @Cacheable
    override fun findCodesOfClass(codeClassId: CodeClassId, languageCode: String): List<Code> = dslContext
        .select(
            ch.difty.scipamato.publ.db.tables.Code.CODE.CODE_CLASS_ID,
            ch.difty.scipamato.publ.db.tables.Code.CODE.CODE_,
            ch.difty.scipamato.publ.db.tables.Code.CODE.LANG_CODE,
            ch.difty.scipamato.publ.db.tables.Code.CODE.NAME,
            ch.difty.scipamato.publ.db.tables.Code.CODE.COMMENT,
            ch.difty.scipamato.publ.db.tables.Code.CODE.SORT
        ).from(ch.difty.scipamato.publ.db.tables.Code.CODE)
        .where(ch.difty.scipamato.publ.db.tables.Code.CODE.LANG_CODE.eq(trimLanguageCode(languageCode))
            .and(ch.difty.scipamato.publ.db.tables.Code.CODE.CODE_CLASS_ID.equal(codeClassId.id)))
        .orderBy(
            ch.difty.scipamato.publ.db.tables.Code.CODE.CODE_CLASS_ID,
            ch.difty.scipamato.publ.db.tables.Code.CODE.SORT
        ).map { r -> Code(r[0] as Int?, r[1] as String?, r[2] as String?, r[3] as String?, r[4] as String?, r[5] as Int) }
}
